keps/sig-apps/20200302-surge-on-daemonsets.md

```diff
@@ -0,0 +1,148 @@
+---
+title: Allow DaemonSets to surge during update like Deployments
+authors:
+  - "@smarterclayton"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-scheduling
+  - sig-node
+reviewers:
+  - "@kow3ns"
+  - "@janetkuo"
+approvers:
+  - "@kow3ns"
+  - "@janetkuo"
+editor: TBD
+creation-date: 2020-03-02
+last-updated: 2020-03-02
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Allow DaemonSets to surge during update like Deployments
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  - [Workload Implications](#workload-implications)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Daemonsets allow two update strategies - OnDelete which only replaces pods when they are deleted and RollingUpdate which supports MinAvailable like Deployments but not Surge. Daemonsets should support Surge in order to minimize DaemonSet downtime on nodes. This will allow daemonset workloads to implement zero-downtime upgrades.
+
+## Motivation
+
+DaemonSets are a key enabler of Kubernetes system-level integrations like CNI, CSI, or per-node functionality. These integrations may have availability impacts on workloads during daemonset updates for a number of reasons, including image pull time or setup. While increasing availability of these daemonsets often requires development investment to manage the handoff between the old instance and the new instance, without the ability to have two pods on the same node these handoffs are complex to implement and typically require higher level orchestration (such as running two daemonsets and round robining updates, or using the OnDelete strategy and orchestrating pod deletes when nodes will be rebooted).
+
+It should be possible for a node level integration to offer zero-downtime upgrades via a DaemonSet without resorting to a higher level orchestration.
+
+### Goals
+
+- Add support for Surge to the DaemonSet rolling update strategy
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
```

Comment:
does PDB need to be updated for this as well?

Reply:
I can add some comments in here, but drain's goal is to remove pods from the node.  If you choose to drain daemonsets, you have to match whatever the PDB constraint was, and since daemonsets ignore unschedulable you don't really get the right outcome from a PDB.  You drain only if the requisite PDB reqt is met, but then immediately after drain succeeds the pod comes right back and PDB goes back up, so a parallel drain is going to potentially see you fail.  If you have a large drain then yes PDB would limit you (as long as the node objects were still there).

Reply:
added implications section